### DISCUSSION TITLE
React-SlideShow WebPack Configuration [Not working]

### DISCUSSION THREAD

**Discussion by EgamoPi** [vercel/next.js#12484] [Help] [answered]
Good morning y'all ! 
Recently started using Next.Js for professional use, I am currently developing a web app for a medium sized african hospital located in Ghana. I am impressed on how easy and efficient I coded 80% of the web app. 
However I am still having a couple of issues with how the next.config.js file works. 
I am trying to implement the react-slideshow library in my app. (https://github.com/femioladeji/react-slideshow), I followed their instructions and they even give the code to add in the next.config.js. 
However once I tried running the app, everything just crashed. 
And they strange thing is that the error output is not even related to the library but with something totally different.

Output from console : 

`(base) ~/.../webApps/aneeja-app >>> yarn dev                                                                                                                                         [130]
yarn run v1.22.4
$ next dev
[ wait ]  starting the development server ...
[ info ]  waiting on http://localhost:3000 ...
[ ready ] compiled successfully - ready on http://localhost:3000
[ event ] build page: /
[ wait ]  compiling ...
[ error ] ./aneeja_assets/medias/instagram.png 1:5
Module parse failed: Unexpected character '' (1:5)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
[ event ] build page: /next/dist/pages/_error
[ wait ]  compiling ...
[ error ] ./aneeja_assets/medias/instagram.png 1:5
Module parse failed: Unexpected character '' (1:5)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
ModuleParseError: Module parse failed: Unexpected character '' (1:5)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
    at handleParseError (/home/sekonoj/Documents/webApps/aneeja-app/node_modules/webpack/lib/NormalModule.js:469:19)
    at /home/sekonoj/Documents/webApps/aneeja-app/node_modules/webpack/lib/NormalModule.js:503:5
    at /home/sekonoj/Documents/webApps/aneeja-app/node_modules/webpack/lib/NormalModule.js:358:12
    at /home/sekonoj/Documents/webApps/aneeja-app/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/home/sekonoj/Documents/webApps/aneeja-app/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at /home/sekonoj/Documents/webApps/aneeja-app/node_modules/loader-runner/lib/LoaderRunner.js:205:4
    at /home/sekonoj/Documents/webApps/aneeja-app/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:85:15
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
[ error ] ./aneeja_assets/medias/instagram.png 1:5
Module parse failed: Unexpected character '' (1:5)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
[ ready ] compiled successfully - ready on http://localhost:3000`

Output from browser : 

`./aneeja_assets/medias/instagram.png 1:5
Module parse failed: Unexpected character '' (1:5)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)`

This is my next.config.js file : 

`
const isProd = process.env.NODE_ENV === 'production'
const withImages = require('next-images')
module.exports = withImages({})

const withCSS = require("@zeit/next-css");
const resolve = require("resolve");
global.navigator = () => null;
module.exports = withCSS({
  webpack(config, options) {
    const { dir, isServer } = options;
    config.externals = [];
    if (isServer) {
      config.externals.push((context, request, callback) => {
        resolve(
          request,
          { basedir: dir, preserveSymlinks: true },
          (err, res) => {
            if (err) {
              return callback();
            }
            if (
              res.match(/node_modules[/\\].*\.css/) &&
              !res.match(/node_modules[/\\]webpack/) &&
              !res.match(/node_modules[/\\]react-slideshow-image/)
            ) {
              return callback(null, `commonjs ${request}`);
            }
            callback();
          }
        );
      });
    }
    return config;
  }
});
`

Before adding the config for the react-slideShow, everything was working perfectly, I'm just totally lost, and I have to deliver this app by the end of this week and I am starting to panic.. It is the first job that I got since the beginning of the quarantine. 

I hope that I was enough clear. Please feel free to ask me any other questions, I will answer in the most detailed way for you Ladies and Gentlemen to be able to help me! Thank you so much, and please... Stay Safe ! <3

Those are other files : 

package.json : 

`{
  "name": "aneeja-app",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@zeit/next-css": "^1.0.1",
    "global": "^4.4.0",
    "next": "9.3.0",
    "next-images": "^1.4.0",
    "react": "16.13.0",
    "react-dom": "16.13.0",
    "react-slideshow-image": "^1.4.2",
    "styled-components": "^5.0.1"
  }
}
`

**Comment by rafaelalmeidatk** [comment, marked as answer]
The problem seems to be with the images import, by default you can't do this since there is no loader defined to handle them (this is why you see the `"You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file"` message).

You can fix that by using something like [`next-optimized-images`](https://github.com/cyrilwanner/next-optimized-images) or [`next-images`](https://github.com/twopluszero/next-images). You can also define the loaders youself with [`url-loader`](https://github.com/webpack-contrib/url-loader) and [`file-loader`](https://github.com/webpack-contrib/file-loader) but it will take more work to get it right.

Alternatively, instead of importing the images in your code, you can paste them in the [`public`](https://nextjs.org/docs/basic-features/static-file-serving) folder and reference them by the `src` attribute like this:

```js
<img src="/my-image.png" alt="my image" />
```

Edit: I reread the code and noticed you already installed `next-images` but the code is overriding `module.exports` (it has two lines defining it) so it isn't being applied. Remove the `module.exports = withImages({})` line and wrap the config: `withImages(withCSS({ ... }))`

**Reply by EgamoPi** [reply]
Thank you for answering that fast! Things are moving, not solved yet but moving ! :) 
The thing is that now I no more know if it's because of the next.config.js file or something else. I think it is the config file since i never had this problem before. 

Here is what I am getting : 

`Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app`

The thing is that I am not even using Hooks and all my components are functional. I am sharing the code where I use the slider library : 

![Screenshot from 2020-05-04 21-06-47](https://user-images.githubusercontent.com/26862670/81013826-57813b80-8e4b-11ea-902c-b0159859e294.png)
![Screenshot from 2020-05-04 21-07-02](https://user-images.githubusercontent.com/26862670/81013845-61a33a00-8e4b-11ea-8ad4-6051d46a36a4.png)

Thank you so much again for helping me, and thank you for showing me how to wrap multiple configs !

Edit : And thank you for giving me different ways of dealing with images. I will check all that later, I want to finish first the "big" of the app.. I still have to figure out how to change fonts or use the google font cdn .. :) 
I probably have to read one more time the documentation of next.js !

**Reply by rafaelalmeidatk** [reply]
Can you show me your package.json? If it is long you may want to share it with https://pastebin.com or similar apps.

By the way, is there a specific reason you aren't using [JSX](https://reactjs.org/docs/introducing-jsx.html) to build your components?

**Reply by EgamoPi** [reply]
Good morning Rafael ! 
Thank you again for you help. 
Here is my package.json 

`{
  "name": "aneeja-app",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@zeit/next-css": "^1.0.1",
    "global": "^4.4.0",
    "next": "9.3.0",
    "next-images": "^1.4.0",
    "react": "16.13.0",
    "react-dom": "16.13.0",
    "react-slideshow-image": "^1.4.2",
    "styled-components": "^5.0.1"
  }
}
`
I checked why I have such messages related with hooks especially when I am not using them, so I checked if I did not have multiple react-dom in my project, not the case. My react version is not that old.. and yeah. 

And to answer your question about the jsxless, my "mentor" thought me that way, and after practicing I realized that I was more productive, more organized with my code without jsx than with. At the end I have less lines of code, and I also think that compilation time decreases. I might be wrong ahah !

**Reply by EgamoPi** [reply]
Hey Rafael ! Hope you're doing well, I just wanted to know if you need any other information about the configs in order to help me with this issue. I am still clueless.. 
Have a good day !

**Reply by rafaelalmeidatk** [reply]
I tried adding this library to a new project and it worked, I might need a reproduction in a repo to take a closer look.

Just to make sure: are you starting the dev server with `yarn dev`? Asking this because I have seen a similar problem with this same error message where the user was starting the server with `next dev` (in the terminal), that was using a global installed version of `next` instead of using the local one

**Reply by EgamoPi** [reply]
Yes, I am starting the dev server with `yarn dev` ! 
I will read on how to do a reproduction and will let you know ! It is really strange, I feel that it is a stupid mistake of my side,  but I am so relieved that the library works for you. Giving me back hope for the deadline ahah! 
Thank you Rafael.

**Reply by EgamoPi** [reply]
Again hello, did you receive the invitation to the repo ?

**Reply by rafaelalmeidatk** [reply]
Hey, sorry for the delay, I got busy with work. I tried the project and the issue is with the config in `next.config.js`, it has some legacy code that is breaking React hooks (changing the `config.externals` array I think).

You can use `next-transpile-modules` to transpile the library that is importing CSS, so this is the fixed version of `next.config.js`:

```js
const withImages = require("next-images");
const withCSS = require("@zeit/next-css");
const withTM = require("next-transpile-modules")(["react-slideshow-image"]);

module.exports = withTM(withImages(withCSS({})));
```

**Reply by EgamoPi** [reply]
Thank you so much, it worked ! Should I report to the slideshow library that their config for next.config.js is outdated and that it can be done with only one line of code ?? :D I love how funny coding can be sometimes, thank you so much for the efforts! Once the web app is done, i'll let you know :)
Have a nice day !

**Comment by EgamoPi** [comment]
@rafaelalmeidatk Good Morning, hope you are doing well! 
I am having a problem related to this discussion, it seems the the slide-show library is causing some issues.. 

I get that after yarn dev : 

`[ wait ]  starting the development server ...
[ info ]  waiting on http://localhost:3000 ...
> Using "webpackDevMiddleware" config function defined in next.config.js.
[ ready ] compiled successfully - ready on http://localhost:3000
[ event ] build page: /next/dist/pages/_error
[ wait ]  compiling ...
[ ready ] compiled successfully - ready on http://localhost:3000
[ event ] build page: /
[ wait ]  compiling ...
[ ready ] compiled successfully - ready on http://localhost:3000
ReferenceError: window is not defined
    at Object../node_modules/react-slideshow-image/dist/react-slideshow-image.min.js (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:2482:3)
    at __webpack_require__ (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:23:31)
    at Module../components/Header.js (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:306:79)
    at __webpack_require__ (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:23:31)
    at Module../pages/index.js (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:4153:76)
    at __webpack_require__ (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:23:31)
    at Object.4 (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:4245:18)
    at __webpack_require__ (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:23:31)
    at /home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:91:18
    at Object.<anonymous> (/home/sekonoj/Documents/webApps/aneeja-app/.next/server/static/development/pages/index.js:94:10)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)`

I do not understand the 'window is not defined'

Thank you for your help!)

**Reply by EgamoPi** [reply]
This is my current next.config.js 

`const isProd = process.env.NODE_ENV === 'production'



const withImages = require('next-images')
const withCSS = require("@zeit/next-css");
const withTM = require("next-transpile-modules")(["react-slideshow-image"]);
const withFonts = require("next-fonts")
const resolve = require("resolve");

global.navigator = () => null;
module.exports = withFonts(withTM(withImages(withCSS({})))); `
